### DISCUSSION TITLE
Can’t instantiate `ViewState` from a `TestStore` when constructed using `BindingViewStore`

### DISCUSSION THREAD

**Discussion by rhysm94** [pointfreeco/swift-composable-architecture#2381] [Ideas]
When writing tests, it’s possible to test `ViewState` if it is constructed directly from your feature state. e.g.
```swift
// Inside Feature
struct State: Equatable {
  var text = “”
  var isLoadingOneThing = false
  var isLoadingAnotherThing = false
}

// Inside View
struct ViewState: Equatable {
  var text: String
  var isLoading: Bool

  init(state: Feature.State) {
    self.text = text
    self.isLoading = state.isLoadingOneThing || state.isLoadingAnotherThing
  }
}
```

When writing tests, it’s easy to assert on your `ViewState`:

```swift
await testStore.send(.someAction) {
  $0.isLoadingOneThing = true
}

XCTAssertTrue(ViewState(testStore.state).isLoading))
```

However, when using the new `Binding` features of TCA along with `ViewState`, if the `ViewState` is initialised using a `BindingViewStore`, it’s no longer possible to write tests against the `ViewState`.
e.g.

```swift
// Inside Feature
struct State: Equatable {
  @BindingState var text = “”
  var isLoadingOneThing = false
  var isLoadingAnotherThing = false
}

// Inside View
struct ViewState: Equatable {
  @BindingViewState var text: String
  var isLoading: Bool

  init(store: BindingViewStore<Feature.State>) {
    self._text = store.$text
    self.isLoading = state.isLoadingOneThing || state.isLoadingAnotherThing
  }
}
```

I can’t find a way to derive or construct a `BindingViewStore` from the `TestStore`, so it’s no longer possible to write the above test code, as we can’t construct our `ViewStore` any longer.

I tried forking TCA and writing an extension on `TestStore where Action: BindableAction` to write something which exposed a `BindingViewStore`, but I couldn’t quite get the types to line up for the `Action` type, so I think I was doing something wrong there 😅

Is there any chance that this is something that could be added, or are there any recommendations on alternative approaches? Ideally I’d like to avoid adding computed properties to my state *and* then assign them in the `ViewState`, but I guess this approach could work as a fallback.

Thanks!

**Comment by stephencelis** [comment]
@rhysm94 Thanks for starting the discussion. This is definitely a gap that ought to be filled, but as we haven't spent a ton of time testing view state transformations we don't know if our API design would be as informed. Do you have any suggestions? How would you like to test your view state?

The simplest solution I could imagine is adding a `bindingViewStore` property, or maybe simply `bindings`, or some other name, to `TestStore`:

```swift
let store = TestStore(…)

await store(.someAction) {
  $0.isLoadingOneThing = true
}
XCTAssertTrue(ViewState(store.bindings).isLoading))
```

We've also talked about adding an `XCTAssertDifference` helper to Custom Dump, so you could maybe even someday get some exhaustivity (in a bit of a nested soup):

```swift
await XCTAssertDifference(store.bindings) {
  await store(.someAction) {
    $0.isLoadingOneThing = true
  }
} changes: {
  $0.isLoading = true
}
```

Any other ideas come to mind?

**Reply by rhysm94** [reply]
@stephencelis That `bindingViewStore` extension (or `bindings`!) is exactly what I envisioned and tried implementing 😀 

I think I just got a bit stuck trying to work out how to transform the `Action` type.

I like the `XCTAssertDifference` as well, for exhaustivity!

**Reply by stephencelis** [reply]
@rhysm94 We have the basic tool sketched out here: https://github.com/pointfreeco/swift-composable-architecture/pull/2394

Wanna take it for a spin and report back?

**Reply by rhysm94** [reply]
@stephencelis Hey, sorry for not responding sooner! I tested this out, and it works perfectly 😁
Exactly what I needed, thank you!
